samples/wifi/provisioning/common/wifi_config.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <string.h>
+#include <zephyr/zephyr.h>
+#include <zephyr/settings/settings.h>
+
+#include "wifi_provisioning.h"
+#include "wifi_prov_internal.h"
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(wifi_prov, CONFIG_WIFI_PROVISIONING_LOG_LEVEL);
+
+#define WIFI_CONFIG_KEY   "wifi_prov/wifi_config"
+
+K_MUTEX_DEFINE(prov_lock);
+static bool prov_state;
+static struct wifi_config config;
+static bool save_in_ram_only;
+
+struct direct_immediate_value {
+	size_t len;
+	void *dest;
+	uint8_t fetched;
+};
+
+static int direct_loader_immediate_value(const char *name, size_t len,
+					 settings_read_cb read_cb, void *cb_arg,
+					 void *param)
+{
+	const char *next;
+	size_t name_len;
+	int rc;
+	struct direct_immediate_value *one_value =
+					(struct direct_immediate_value *)param;
+
+	name_len = settings_name_next(name, &next);
+
+	if (name_len == 0) {
+		if (len == one_value->len) {
+			rc = read_cb(cb_arg, one_value->dest, len);
+			if (rc >= 0) {
+				one_value->fetched = 1;
+				LOG_DBG("immediate load: OK.");
+				return 0;
+			}
+
+			LOG_DBG("fail (err %d).", rc);
+			return rc;
+		}
+		return -EINVAL;
+	}
+
+	/* other keys aren't served by the callback
+	 * Return success in order to skip them
+	 * and keep storage processing.
+	 */
+	return 0;
+}
+
+static int load_immediate_value(const char *name, void *dest, size_t len)
+{
+	int rc;
+	struct direct_immediate_value dov;
+
+	dov.fetched = 0;
+	dov.len = len;
+	dov.dest = dest;
+
+	rc = settings_load_subtree_direct(name, direct_loader_immediate_value,
+					  (void *)&dov);
+	if (rc == 0) {
+		if (!dov.fetched) {
+			rc = -ENOENT;
+		}
+	}
+
+	return rc;
+}
+
+static int wifi_config_save(struct wifi_config *config)
+{
+	int rc;
+
+	rc = settings_save_one(WIFI_CONFIG_KEY, config, sizeof(struct wifi_config));
+	if (rc != 0) {
+		LOG_WRN("Error! Cannot write condig.");
+		return -EIO;
+	}
+	return 0;
+}
+
+static int wifi_config_read(struct wifi_config *config)
+{
+	int rc;
+
+	rc = load_immediate_value(WIFI_CONFIG_KEY, config, sizeof(struct wifi_config));
+	if (rc != 0) {
+		LOG_WRN("Error! Cannot read config.");
+		return -EIO;
+	}
+	return 0;
+}
+
+static int wifi_config_delete(void)
+{
+	int rc;
+
+	rc = settings_delete(WIFI_CONFIG_KEY);
+	if (rc != 0) {
+		LOG_WRN("Error! Cannot delete config.");
+		return -EIO;
+	}
+	return 0;
+}
+
+static int wifi_settings_init(void)
+{
+	int err;
+
+	err = settings_subsys_init();
+	if (err) {
+		LOG_WRN("settings_subsys_init failed (err %d).", err);
+		return err;
+	}
+
+	return 0;
+}
+
+bool wifi_has_config(void)
+{
+	bool has_config = false;
+
+	if (k_mutex_lock(&prov_lock, K_NO_WAIT) == 0) {
```

Comment:
I'm doubting in sense of multithread access protection for this check (it is `bool` type). State might change right after this function returns anyway.

Reply:
I think the use of mutex is correct, which is copied from zephyr's doc. If the mutex is locked successfully, the thread owns the mutex even before the `k_mutex_lock` returns.

Reply:
Why `K_NO_WAIT` though? Woudn't it be more user friendly approach to actually wait for the mutex to be unlocked instead of returning an error (or even potentially incorrect information in this particular function)?

Reply:
You're right. Let the caller wait for certain time to lock the mutex, instead of returning an error immediately. I will add some waiting time.